canonical-web-and-design/vanilla-framework
# POC: replace example-js with codepen embeds

## Done

Proof of concept to explore replacing examples with docs with CodePen embeds.

## QA

- Pull code
- Run `./run serve --watch`
- Open http://0.0.0.0:8101/ or demo https://vanilla-framework-canonical-web-and-design-pr-2702.run.demo.haus/
- All examples in the docs should be replaced with equivalent CodePen embeds

## Screenshots
<img width="979" alt="Screenshot 2019-12-11 at 11 41 57" src="https://user-images.githubusercontent.com/83575/70614751-600e7680-1c0b-11ea-9c22-d042f163d620.png">

<img width="986" alt="Screenshot 2019-12-11 at 12 03 20" src="https://user-images.githubusercontent.com/83575/70616242-40c51880-1c0e-11ea-89ee-850aa6ed6184.png">

## Details

### Pros of using CodePen

- we can simplify example-js as we won't need to build iframe and code views (especially when we want to include JS tab, maybe React in future)
  - simplified example-js is still needed, because we need to fetch and parse example HTML and send it to codepen embed
- we can easily add JS code tab without implementing it ourselves
  - possibly in future we can use CodePen for React examples
- codepen is known and respected tool in webdev community
- free 'Open in CodePen' button that makes it super easy to open any example in CodePen and change/update it to play with vanilla patterns
  - would also make super easy to create small test case examples for issues reporting
- built in syntax highlighting for code
- built in support for preprocessors for HTML (like HAML etc) CSS (SCSS, etc) JS (Babel) 
- build in x1,x0.5,x0.25 scaling option, to make it easier to test responsive components

### Cons of CodePen (vs custom solution)

- we can't automatically adjust height of the embed like we do with iframe
  - we can define the height of the embed we want
  - if we want per-example adjustment, we would need to store height somewhere along the example (some metadata) and use it when creating the embed
- we have limited control on styling (some theming possible in terms of adjusting colors)
  - see below for theming details
- we have limited control on layout of the embed - for example code (HTML/JS) can be only on a side of 'Result', we can't show code under the rendered example like we do currently
- we rely on external resource to render our examples
  - because examples are rendered on codepen it may take longer for them to load
- no 'copy to clipboard' button
- need to inject some margin on body to prevent examples from sticking to the borders

## Themeimg

There are 2 built in themes dark and light:

<img width="886" alt="Screenshot 2019-12-11 at 12 10 26" src="https://user-images.githubusercontent.com/83575/70617081-1a07e180-1c10-11ea-8b89-79d9a02112a5.png">

<img width="869" alt="Screenshot 2019-12-11 at 12 09 32" src="https://user-images.githubusercontent.com/83575/70617080-196f4b00-1c10-11ea-8d4d-e9cee3364549.png">

If we want more adjustment having an account on CodePen allows to create one custom theme (we would probably want to have vanilla account in such case):

<img width="872" alt="Screenshot 2019-12-11 at 12 14 01" src="https://user-images.githubusercontent.com/83575/70617168-47548f80-1c10-11ea-9e67-9735fd03154b.png">

Theming options are quite limited, they mostly allow to choose colors for different parts of the UI (inlcuding choosing one of preset syntax highlighting themes).
PRO (paid) users have more control with custom CSS.

<img width="1353" alt="Screenshot 2019-12-11 at 12 10 55" src="https://user-images.githubusercontent.com/83575/70617226-6d7a2f80-1c10-11ea-8a23-9e6f917e6083.png">


## Docs

https://blog.codepen.io/documentation/prefill-embeds/
https://blog.codepen.io/documentation/pro-features/unlimited-embed-themes/
